### PR TITLE
Readme for Setting up External Idp

### PR DESCRIPTION
**Pull request by TejasRGitHub**
### Feature or Bugfix

- Feature

### Detail

Readme instruction for setting up external Idp which supports OIDC in data.all 

### Relates
- [<URL or Ticket>](https://github.com/awslabs/aws-dataall/issues/872)

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)? N/A
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization? N/A
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features? N/A
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored? N/A
- Are you introducing any new policies/roles/users? N/A
  - Have you used the least-privilege principle? How?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pages/architecture.md [hunk 1]
```diff
@@ -105,6 +105,11 @@ abstracting the different IdP providers protocols.
 **Note**: data.all doesn't have a user store and does not create or manage groups.
 It relies only on information provided by the IdP; such as username, email, groups, etc...
 
+Data.all can also be configured to use any external Idp which uses OIDC. With this , data.all user will be authenticated
+with your Idp. With external Idp, data.all doesn't maintain the user groups information. In order for data.all modules to work seamlessly, you will have to implement methods provided in the
+dataall/base/services/ServiceProvider.py and provide an instance of your implementation in the dataall/base/services/ServiceProviderFactory.py.
+For more information on deploying with external Idp, please checkoout the [Deploy to AWS](./deploy/deploy_aws.md) section.
+
```
Comment (dlpzx) [inline pages/architecture.md:112]: typo: "checkout"

File: pages/architecture.md [hunk 2]
```diff
@@ -105,6 +105,11 @@ abstracting the different IdP providers protocols.
 **Note**: data.all doesn't have a user store and does not create or manage groups.
 It relies only on information provided by the IdP; such as username, email, groups, etc...
 
+Data.all can also be configured to use any external Idp which uses OIDC. With this , data.all user will be authenticated
+with your Idp. With external Idp, data.all doesn't maintain the user groups information. In order for data.all modules to work seamlessly, you will have to implement methods provided in the
```
Comment (dlpzx) [inline pages/architecture.md:109]: typo: space before comma "With this"

File: pages/architecture.md [hunk 3]
```diff
@@ -207,6 +214,8 @@ As explained in the frontend section, Amazon Cognito is used for Authentication
 In Amazon API Gateway we again use [Cognito for Authorization](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-integrate-with-cognito.html). 
 With an Amazon Cognito user pool, we control who can access our GraphQL API. 
 
+**Note** - You can also deploy data.all with your own IDP. In order to do so please follow instruction in ["Deploy to AWS"](./deploy/deploy_aws.md) section
```
Comment (dlpzx) [inline pages/architecture.md:217]: The general recommendation in "deploy data.all with your own IdP" is to keep using Cognito as a proxy. So in this paragraph I would be a bit more specific: "You can also replace Cognito authorization by a custom authorizer using your own IdP"

File: pages/deploy/deploy_aws.md [hunk 4]
```diff
@@ -197,41 +210,42 @@ have listed and defined all the parameters of the cdk.json file. If you still ha
...
+| reauth_config                                 | Optional              | A dictionary containing a list of API operations that require a user to re-authenticate before proceedind (`reauth_apis`) and a time to live (`ttl`) for how long a user's re-auth session is valid to perform re-auth APIs before having to re-authenticate again                                                                                      |
+| custom_auth                                   | Optional              | A dictionary containing set of parameters to setup External Idp Authentication in data.all. Custom Auth Configuration : `provider`, `url`, `redirect_url`, `client_id`, `response_types`, `scopes`, `jwks_url`, `claims_mapping` (Nested dictionary containing configuration : `user_id`, `email`). All of the configuration are required if setting data.all with an external OIDC supported Idp|
```
Comment (dlpzx) [inline pages/deploy/deploy_aws.md:248]: very minor: Authentication and Authorization

File: pages/deploy/deploy_aws.md [hunk 5]
```diff
@@ -315,6 +329,85 @@ deploy to 2 deployments accounts.
   }
 }
 ```
+
+**Example 3** - Deployment with Custom Authentication
```
Comment (dlpzx) [inline pages/deploy/deploy_aws.md:333]: Love this!!!

File: pages/deploy/deploy_aws.md [hunk 6]
```diff
@@ -315,6 +329,85 @@ deploy to 2 deployments accounts.
...
+    ]
+  }
+}
+```
+
+### 6.1 (Important) - Configure data.all to use external Idp
+
+Please read this section only if you are deploying data.all with an external Idp.
+
+Out of the box, data.all uses Cognito as the Idp and the user pool provider. Data.all can also be configured to use an external Idp
```
Comment (dlpzx) [inline pages/deploy/deploy_aws.md:381]: I would first mention that Cognito can be used as a proxy and can be integrated with other IdPs. But for the case where a customer wants to fully replace Cognito then it is possible thanks to "custom_auth" configuration

File: pages/deploy/deploy_aws.md [hunk 7]
```diff
@@ -315,6 +329,85 @@ deploy to 2 deployments accounts.
...
+    def get_groups_for_user(self, user_id):
+        # Your implementation
+    
+    def list_groups(self, envname: str, region: str):
+        # Your implementation
+```
+
+Once you implement these methods, create an instance of your provider in dataall/base/service/ServiceProviderFactory.py. Please follow the 
+instructions in the comments to create and return and instance of your custom service provider.
+
```
Comment (dlpzx) [inline pages/deploy/deploy_aws.md:409]: Maybe we can add here a small code snippet example

File: pages/architecture.md (hunk 2 above)
Comment (TejasRGitHub) [reply pages/architecture.md:109]: Addressed

File: pages/architecture.md (hunk 3 above)
Comment (TejasRGitHub) [reply pages/architecture.md:217]: Added a statement which gives clear directions to the user when they want to implement a custom authorizer. Let me know if that sounds fine.

File: pages/deploy/deploy_aws.md (hunk 4 above)
Comment (TejasRGitHub) [reply pages/deploy/deploy_aws.md:248]: Addressed.

File: pages/deploy/deploy_aws.md (hunk 6 above)
Comment (TejasRGitHub) [reply pages/deploy/deploy_aws.md:381]: Addressed this. Please take a look

File: pages/deploy/deploy_aws.md (hunk 7 above)
Comment (TejasRGitHub) [reply pages/deploy/deploy_aws.md:409]: Added . Thanks !

**Review by dlpzx — APPROVED** [review]

**Pull request merged**